### ISSUE TITLE
Redirects v1 (LTS) 

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#7707]
Ghost currently doesn't have a way to add redirects, e.g. if a post URL changes or if you migrate from another platform and you want to add a catch all to modify the permalink structure to something Ghost supports.

On Ghost(Pro) we've always supported this by allowing users to request redirects & applying them one layer above Ghost. For self-hosters, we recommended adding redirects in the layer above Ghost - nginx or apache. As this is possible in most cases, redirects haven't been seen as a high-level priority.

However, this setup has lead to a few problems. First, users get a poor experience. On Ghost(Pro) redirects have to be requested through support, not everyone thinks to ask, and you can't tweak or optimise them. For self-hosters they have to keep track of a critical part of their SEO outside of Ghost, in a highly technical system, when in the majority of cases redirects are a feature used for SEO (marketing). 

Structurally it depends on configuring and restarting the layer above Ghost - nginx or apache. This is fine in the common case of a one-time-setup, but if you want a system where redirects can be added regularly, this becomes problematic. If you run more than just Ghost you have to configure & restart a piece of software that affects all of your sites, if you get it wrong the whole lot can go offline.

Bottom line is that this is far from optimal, and we want to start adding provision for redirects in Ghost itself.

## v1 Solution (LTS)

For now, we're going to be adding a very simple version of this feature to the Ghost LTS stream. 

This version should be file-based. It should be possible to provide a redirects config file in the `content/data` directory. Ghost will read this file when it starts and use an express router to dynamically build the routes which perform the redirects.

- Redirects should happen before anything else.
- Both 301 and 302 redirects should be supported.
    - 301 redirects should have a long max-age so that they are cached by any layers above Ghost, for optimal performance. 
   - 302 redirects should have a max-age of 0 

Redirects contain a from and a to path, and support regular expression matches. This may be a bit tricky, as we need to be clear on what regular expressions we do/don't support, whether they should or shouldn't be escaped, etc.

E.g. I want to redirect from `^/post/[0-9]+/([a-z0-9\-]+)` to `/$1`, this format is I think the one from tumblr? The idea is to match the slug on the end and redirect to just that slug, as we would use it in Ghost.

The start of the regular expression should count from the end of the url provided in `config.url` and it should work regardless of whether the slash is present in either location.

Originally I thought we should restrict redirects to only affecting the blog frontend. However, I can think of legitimate use cases where you may need or want to redirect other routes, so let's build this feature permissively for a change and see what happens 😄 

The trickiest part to figure out, IMO, is the redirect file format. I have 2 competing ideas in mind:

**redirects.json**

Contains an array of objects, each object contains type, from and to.

Example:

```
[
  {
    "type": "301",
    "from":  "^/post/[0-9]+/([a-z0-9\-]+)"
    "to": "/$1"
  },
  {
    "type": "301",
    "from": "/my-old-blog-post/",
    "to": "/revamped-url/" 
  }
]
```

Pros: well understood & standardised data format, easily machine read and writable, works well with JS in Ghost.
Cons: not easily human writeable / readable

**redirects.conf**

Simple text file, each redirect on a new line, a space separates each piece of info in the order `type from to`. Essentially a CSV with a space separator.

```
301 ^/post/[0-9]+/([a-z0-9\-]+) /$1
301 /my-old-blog-post/ /revamped-url/
```

Pros: very easily human read/writable
Cons: slightly trickier to read with code, need to document the format (although it's peasy!)

I am absolutely on the fence about which of these to use. I think it is much easier as a not-necessarily technical human to edit and add to the redirects.conf file, and it might suit our purposes more, however it feels weird to propose not-JSON.

## v2 Solution (later)

I want to also document here, that this solution, although fully-featured, is considered "v1". 

Version 2 will have an interface very similar to the navigation UI in Ghost-Admin, and store the redirects in a new database table. This will be far more usable for the intended audience than a file on a server. The redirects file will be read in the first time after we add this feature, and then not used anymore.

As we are intending on replacing it, this means the decision on the file format for the redirects file doesn't matter too much, and we should do what gets us to a solution quickest.

**Comment by kirrg001** [comment]
I would like to use `redirects.json` format. Some reasonings:
- file format is easier to write for me on Ghost(Pro)
- we will move to V2 soon (file-less redirects), so the decision is not very important

I thought about using "permanent: true|false" instead of "type". So people can define either to redirect permanent or temporary - any opinions?

**Comment by kirrg001** [comment]
- [ ] add documentation when merged #7719

**Comment by kirrg001** [comment]
Closing. I've added the later label, because it could happen that we revisit and implement solution V2 (see issue description).
